golem-worker-service-base/src/service/worker/default.rs

```diff
@@ -418,6 +503,57 @@ where
         Ok(get_json_from_typed_value(&typed_value.result))
     }
 
+    async fn invoke_and_await_parsed_function(
```

Comment:
A few more notes:

There is already a type-checker in the peripherals (component-service/worker-service), before calling worker-executor which we decided last year May. If that's the case, then why is it that having the full parsed-function-name in the same service is a wrong/suboptimal choice now?

Reply:
Let's get back to the context of this PR. Firstly, I already have a parsed-function-name in hand, which can give you `ExportFunction` given function-parameters and function-result types that are also input to this function . This ExportFunction details are used to feed type checkers later on. Inside these checkers, I could also see the name getting parsed again to see if it was an indexed function or not. With ParsedFunctionName (and parameter and result types already fed,  we avoided all these)

In all other functions (or more precisely what only existed before), is  taking function name as a string, we load the metadata again to further resolve export-function and use this data to type check, along with re-parsing these stringified function name again such as to see whether it is indexed. I hope this answers why this new function in this PR taking these as arguments.


So ParsedFunctionName as an input is not a luxury, it's literally to avoid calling the metadata again to resolve the export-function and re-parsing 

Yes invoking the actual function still takes a string function name, but I am not changing all that in this PR.

Reply:
This "type checking" is not really type checking but interpreting the json format. I wrote about this how in the other thread. 

(Also that we decided to do it like this last May is not a good argument, we are continuously learning and improving our design and then we were creating a prototype while now trying to optimise something near production ready)

Reply:
Currently type-annotated-value function parameters are also validated against the metadata. 
Based on the above statement, it does mean we don't require this validation.
But  the output of worker-executor function invocation returns a `Val` which needs to be correlated against the actual data to get the real `Value` (Json or TypeAnnotatedValue) and that requires downloading the metadata regardless of the pathways (worker-bridge, or json parameters through golem-cli)

```scala
#[derive(Clone, PartialEq, ::prost::Message)]
pub struct InvokeResult {
    #[prost(message, repeated, tag = "1")]
    pub result: ::prost::alloc::vec::Vec<::golem_wasm_rpc::protobuf::Val>,
}
```

I think based on your argument above, I can see following four steps:

* `InvokeResult` should be `TypeAnnotatedValue` - that the worker-executor should be responsible for mapping the `Val` against the `Metadata` and return a `TypeAnnotatedValue`. A `TypeAnnotatedValue` can be converted to Json without metadata, and can be done by the worker-service side (to satisfy those who need Json result).

 * Once that is done, then the worker-service can avoid converting the output `Val` to `TypeAnnotatedValue` function parameters (the interpretation of json format) and therefore avoid downloading Metadata for this reason.
That is avoids calling this function:

```scala
 fn validate_function_result(
        output: InvokeResult // output of worker-invocation which is nothing but `Vec<Val>`
        expected_types: Vec<AnalysedFunctionResult> // this comes from metadata download
        calling_convention: CallingConvention,
    ) -> Result<TypeAnnotatedValue, Vec<String>> 
```

 
 * However, we have 1 more metadata-download at worker-service side. This is to validate (or interpret, not sure what to call this) TypeAnnotatedValue function parameters against the component metadata. Now I think this part is definitely not required regardless of this PR - i.e, avoid calling the below function as such and simply forward to worker-executor
 
<img width="906" alt="image" src="https://github.com/golemcloud/golem/assets/7448613/1cdde88d-a0d7-486f-92b8-9632125cbbd3">

* Now this can be debatable/contoversial. Even Json parameters doesn't need to get validated/interpreted against metadata to retrieve a `Val` to call worker-executor. It can be `Vec<serde_json::Value>`  itself and avoid component-metadata download completely. I do believe if worker-executor is handling most of such responsibilities, and it is much beneficial to take this step and do it for Json pathways too. 


Now my final concern which I really hope you take it in the right sense:
Was that the responsibility of this PR? 

If the responsibility of the PR is avoid calling metadata download for Rib, sure I can do, but we are doing that anyway in worker-service. So services that don't use Rib, but use these interfaces end up smashing component service.

Reply:
>(Also that we decided to do it like this last May is not a good argument, we are continuously learning and improving our design and then we were creating a prototype while now trying to optimise something near production ready)

Both of us know my reasoning to keep this is not "we already did this in May, and let's not change it" :D 
I definitely didn't reason it like that. Please excuse if it sounded so.

My point is, we are having system design checks (something what we did in the past and may not be relevant anymore) as part of the smaller tasks/tickets which essentially invalidate the ticket.

Reply:
> It implies the worker-executor should be responsible for mapping the Val against the Metadata and return a TypeAnnotatedValue

This is one possibility, and it allows to keep the current JSON response format, but increases the response payload size unnecessarily for cases when it is not needed. 
There are (at least) two more possibilities to consider:
- if types for the result are only for the JSON format we can potentially change the JSON format itself, but in this case without any field and constructor names (basically `Val` directly serialized to JSON), it would be a very user experience 
- We could make `Val` a bit more self-describing - basically having field/enum/flag/constructor names that would make it enough alone to convert to JSON but is probably a bit less information, but then it would be always coupled with this information even where it's not needed.

So probably your suggestion is the best out of these three.

Note that I think it is a good place to do this mapping in the worker-executor because it is natural to have the component metadata available (as it even has the whole component binary downloaded and cached). 

> This definitely implies, the function-invocation should support accepting type annotated value

I don't understand why it would need that. It currently works by just getting
```protobuf
  repeated wasm.rpc.Val input = 3;
```
so it means we are not sending any type information to it anyway as it already downloads and uses it internally. 
If we somehow have a `TypeAnnotatedValue` (from Rib evaluation?) we can just convert it to `Val` as you wrote. 
The problem is if we have a JSON from the REST API, and what I wrote above (and seems like you also writing it) is that the responsibility of understanding that JSON format can be moved to the worker executor as an optimisation, basically the same way as we had a gRPC endpoint in worker service for a while accepting JSON, during the migration from Scala. (It is deleted now).

Yes it feels more correct to have this mapping from JSON to Val in the worker bridge as it is similar a bit (just very simplified) to evaluating a Rib on a request.

But I think it worths it if it can avoid us running into a very difficult cache invalidation / race condition problem which triggered the whole discussion. 


> If the responsibility of the PR is avoid calling metadata download for Rib, sure I can do, but we are doing that anyway in worker-service. So services that don't use Rib, but use these interfaces end up smashing component service.

I think the original ticket's purpose was not "Rib only" but to solve the repeated metadata download in worker service by caching (no matter which part of worker service). This might have been confusing because we have a confusing naming around worker service (proxy, bridge, api gateway, rib) that we should improve. (We have two "components" merged into one service and some old and new naming and we are quite randomly using them in discussions)

Reply:
Why is caching metadata easy in worker executor and hard in the worker service?

Component metadata is per component _version_ and which one we need depends on the worker:
- a worker gets an associated component version when it is first created, by the worker executor
- it can be updated later using the hot update feature

In would seem like the first one could be moved into the worker service level, so it would always figure out the component version before calling the worker executor. But it's not good because there are cases where a worker is spawn without touching the worker executor, for example when doing a local RPC call within the executor.

The second is even more tricky because even if all hot update request goes through worker service, it is just an _attempt_ to update the worker, which gets enqueued and later tried, and can potentially fail. So the exact time when a worker starts using a new component version cannot be determined only by polling the worker executor.

This is all complicated by the fact that worker service is horizontally scalable and load balanced so requests for the same worker potentially go through different nodes of worker service which all need to properly deal with these invalidation scenarios etc.

Even if we would ignore the above points about moving the "which component version should I use for a new worker" question and somehow move that logic to the worker service, then we have the issue of invalidating this knowledge in our system. Which would require enumerating all worker service nodes and the problem is that cannot be done atomically, and requests to the same worker are potentially arriving into different worker service nodes because it is load balanced.

Now compare all this with just doing this on the worker executor level:

- it is sharded - if we cache something for a worker that's always in the same executor (until rebalancing) 
- it is already caching whole wasmtime components (a result of downloading and compiling a component from component service)
- it knows exactly when an update succeeds and can update its internal caches - and no need to do any cross-node invalidation because of it being sharded

Reply:
It offers a better reliable developer experience, there must be some place where we can do type checking, validation, resolving references; pre-compilation, validating references to functions, etc.

Say this for example:

`let x = foo(1)` is my Rib expression. What exactly is `foo` here? A variant or a function call? Rib doesn't know it, unless it knows about the type details of the component user is referring to?

Another example

`let x = my_exported_function(1, {})`. Is `{}` an empty flag or an empty record ? Both flag and record starts and ends with a curly braces. The answer depends on the function parameter types which Rib doesn't know without downloading the metadata. 

Anyway this explanation doesn't mean I want this download again.
I think for performance benefits, which is super important, I am removing these downloads of metadata at Rib side.

And as we discussed, we may need to do the same for some other functionalities of worker-service which you explained in detail to my proposal/steps.

Reply:
> It offers a better reliable developer experience, there must be some place where we can do type checking, validation, resolving references; pre-compilation, validating references to functions, etc.

But all these things doesn't make sense to me in this context of _invocation_

When you develop the rib expression you don't have a particular worker you are invoking. So this validation at the time of invocation doesn't help with it at all